llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -288,6 +293,89 @@ bool AIE2PreLegalizerCombinerImpl::tryToCombineVectorInserts(
   return true;
 }
 
+/// \returns true if it is possible to combine the below sequence of MIRs
+/// From : %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem16.I512),
+/// %2(<32 x s16>), %0(s32), %1(s32)
+///        %20:_(s16) = G_TRUNC %10(s32)
+///        %30:_(s20) = G_SEXT %20(s16)
+/// To :   %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem16.I512),
+/// %2(<32 x s16>), %0(s32), %1(s32)
+///        %30:_(s20) = G_TRUNC %10(s32)
+/// Or even:
+/// From : %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem8.I512),
+/// %2(<64 x s8>), %0(s32), %1(s32)
+///        %20:_(s8) = G_TRUNC %10(s32)
+///        %30:_(s20) = G_SEXT %20(s8)
+/// To :   %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem8.I512),
+/// %2(<64 x s8>), %0(s32), %1(s32)
+///        %30:_(s20) = G_TRUNC %10(s32)
+/// This also enables S20Narrowing for vextract
+bool AIE2PreLegalizerCombinerImpl::tryToCombineTruncExt(
+    Register DstReg, bool SignVal, unsigned SrcEltSize) const {
+  // Checks if a given register has non-debug user with a specific opcode and
+  // destination size, and return that user.
+  auto GetUseWithOpCode =
+      [&](const Register Reg, const unsigned OpcodeToCheck,
+          const unsigned DstSize) -> std::optional<MachineInstr *> {
+    for (auto &Use : MRI.use_nodbg_instructions(Reg)) {
+      if (Use.getOpcode() == OpcodeToCheck) {
+        const LLT DstRegTy = MRI.getType(Use.getOperand(0).getReg());
+        if (DstRegTy.getSizeInBits() == DstSize)
+          return &Use;
+      }
+    }
+    return std::nullopt;
+  };
+
+  if (auto Trunc =
+          GetUseWithOpCode(DstReg, TargetOpcode::G_TRUNC, SrcEltSize)) {
+    MachineInstr *TruncMI = Trunc.value();
+    const unsigned ExtOpcode =
+        SignVal ? TargetOpcode::G_SEXT : TargetOpcode::G_ZEXT;
+    const Register UseDstReg = TruncMI->getOperand(0).getReg();
+    // Ensure G_TRUNC has a single non-debug user before safely eliminating it.
+    if (!MRI.hasOneNonDBGUser(UseDstReg))
+      return false;
+    if (auto Ext = GetUseWithOpCode(UseDstReg, ExtOpcode, 20)) {
+      MachineInstr *ExtMI = Ext.value();
+      MachineIRBuilder MIRBuilder(*ExtMI);
+      MIRBuilder.buildInstr(TargetOpcode::G_TRUNC, {ExtMI->getOperand(0)},
+                            {DstReg});
+      ExtMI->eraseFromParent();
+      TruncMI->eraseFromParent();
+      return true;
+    }
+  }
+  return false;
+}
+
+bool AIE2PreLegalizerCombinerImpl::tryToCombineVExtractElt(
+    MachineInstr &MI) const {
+  const Register DstReg = MI.getOperand(0).getReg();
+  // In this case of G_INTRINSIC operand 1 is target intrinsic
+  const Register SrcReg = MI.getOperand(2).getReg();
+  const Register IdxReg = MI.getOperand(3).getReg();
+  const Register SignReg = MI.getOperand(4).getReg();
+
+  const auto SignVal = getIConstantVRegSExtVal(SignReg, MRI);
+  if (!SignVal)
```

Comment:
This looks like an assert. We always need constant signal to be able to handle this intrinsic.

Reply:
I thought so too, but actually we can have `VEXTRACT` with a dynamic sign coming from `$crVaddSign`